### PR TITLE
Estimate cost of inequality comparision filters

### PR DESCRIPTION
**Pull request by raunaqmorarka**
On top of https://github.com/trinodb/trino/pull/11469

## Description

When there is some overlap between the ranges of the
operands of an inequality expression, we approximate
it's cost assuming uniform distribution of values
within each range and that all pairs of distinct values
from both ranges exist.

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

CBO
> How would you describe this change to a non-technical end user or system administrator?

Improves query plans in the presence of inequality expressions

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Improve query plans in the presence of inequality expressions. ({issue}`11518`)
```

### CODE REVIEW

**Comment by raunaqmorarka** [comment]
TPC benchmarks
[Inequality estimation sf1000 orc unpartitioned.pdf](https://github.com/trinodb/trino/files/8281840/Inequality.estimation.sf1000.orc.unpartitioned.pdf)

[Inequality estimation sf1000 orc partitioned.pdf](https://github.com/trinodb/trino/files/8281841/Inequality.estimation.sf1000.orc.partitioned.pdf)

File: core/trino-main/src/main/java/io/trino/cost/FilterProjectAggregationStatsRule.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    protected Optional<PlanNodeStatsEstimate> doCalculate(FilterNode node, StatsProvider sourceStats, Lookup lookup, Session session, TypeProvider types)
+    {
+        if (!isUnestimatedFilterCostApproximationEnabled(session)) {
+            return Optional.empty();
+        }
+        PlanNode nodeSource = resolveGroup(lookup, node.getSource());
+        AggregationNode aggregationNode;
+        if (nodeSource instanceof ProjectNode) {
+            ProjectNode projectNode = (ProjectNode) nodeSource;
+            if (!projectNode.isIdentity()) {
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/cost/FilterProjectAggregationStatsRule.java:68]: why only identify?

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 2]
```diff
@@ -239,4 +249,61 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+
+        switch (operator) {
+            case LESS_THAN:
+            case LESS_THAN_OR_EQUAL:
+                // left is always lesser than right
+                if (leftExpressionRange.getHigh() < rightExpressionRange.getLow()) {
+                    PlanNodeStatsEstimate.Builder estimate = PlanNodeStatsEstimate.buildFrom(inputStatistics);
+                    leftExpressionSymbol.ifPresent(symbol -> estimate.addSymbolStatistics(symbol, leftExpressionStatistics));
+                    rightExpressionSymbol.ifPresent(symbol -> estimate.addSymbolStatistics(symbol, rightExpressionStatistics));
+                    return estimate.build();
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:280]: consider extracting lines 277-280 to a method and reuse it in `GREATER_THAN_OR_EQUAL` case.

**Review by lukasz-stec — APPROVED** [review]
lgtm

File: core/trino-main/src/main/java/io/trino/cost/FilterProjectAggregationStatsRule.java (hunk 1 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/FilterProjectAggregationStatsRule.java:68]: I was trying to match, as closely as I could, for the cases where we don't know the estimate due to lack of aggregate symbol stats. For non-identity projections I wasn't sure if lack of estimate could be due to an unestimated projection. We also don't have a non-indentity projection case in TPC benchmarks.

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 3]
```diff
@@ -239,4 +249,61 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+        StatisticRange rightExpressionRange = StatisticRange.from(rightExpressionStatistics);
+        if (leftExpressionRange.isEmpty() || rightExpressionRange.isEmpty()) {
+            return PlanNodeStatsEstimate.unknown();
+        }
+
+        switch (operator) {
+            case LESS_THAN:
+            case LESS_THAN_OR_EQUAL:
+                // left is always lesser than right
+                if (leftExpressionRange.getHigh() < rightExpressionRange.getLow()) {
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:276]: we could do that but I'm preferring simplicity over being exactly accurate here

**Comment by sopel39** [comment]
There are some regressions in partitioned `tpcds` (some lesser regressions in unpart tpcds). Why it's the case?

**Comment by raunaqmorarka** [comment]
> There are some regressions in partitioned `tpcds` (some lesser regressions in unpart tpcds). Why it's the case?

The only actual regression is in tpch/q04 partitioned (which we already knew about, it's due to bad NDV estimate), the rest is benchmark flakiness.

**Comment by sopel39** [comment]
> The only actual regression is in tpch/q04 partitioned (which we already knew about, it's due to bad NDV estimate), the rest is benchmark flakiness.

What about tpcds/q93, q95, 97, 78? Did plans change for these?

**Comment by raunaqmorarka** [comment]
> > The only actual regression is in tpch/q04 partitioned (which we already knew about, it's due to bad NDV estimate), the rest is benchmark flakiness.
> 
> What about tpcds/q93, q95, 97, 78? Did plans change for these?

The inequality estimation code path is executed only for q72 in tpcds and there was no change to plan there.

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 4]
```diff
@@ -239,4 +249,61 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+            boolean isUnestimatedFilterCostApproximationEnabled)
+    {
+        if (leftExpressionStatistics.isUnknown() || rightExpressionStatistics.isUnknown()) {
+            return PlanNodeStatsEstimate.unknown();
+        }
+
+        StatisticRange leftExpressionRange = StatisticRange.from(leftExpressionStatistics);
+        StatisticRange rightExpressionRange = StatisticRange.from(rightExpressionStatistics);
+        if (leftExpressionRange.isEmpty() || rightExpressionRange.isEmpty()) {
+            return PlanNodeStatsEstimate.unknown();
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java]: I've updated it to return 0

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 5]
```diff
@@ -239,4 +249,61 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                // left is always lesser than right, no overlap
+                else if (leftExpressionRange.getHigh() < rightExpressionRange.getLow()) {
+                    return inputStatistics.mapOutputRowCount(rowCount -> 0.0);
+                }
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported inequality operator " + operator);
+        }
+        if (isUnestimatedFilterCostApproximationEnabled) {
+            return inputStatistics.mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:305]: There's 2 main cases where inequality estimate is used, both are in lineitem for tpch
commitdate < receiptdate, 0.63 selectivity with left range completely overlapping right range
shipdate < commitdate, 0.49 selectivity with left range overlap percentage with right range is 97%.

Although it won't matter for TPC benchmark, we can tweak the heuristic to something like below
```
double overlappingRangeFraction = leftRange.overlapPercentWith(leftRange.intersect(rightRange));
double alwaysLessRangeFraction = 0.0;
if (leftRange.getLow() < rightRange.getLow()) {
       alwaysLessRangeFraction = leftRange.overlapPercentWith(new StatisticRange(leftRange.getLow(), rightRange.getLow(), NaN));
}
double filterFactor = alwaysLessRangeFraction + (UNKNOWN_FILTER_COEFFICIENT * overlappingRangeFraction);
```

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 6]
```diff
@@ -239,4 +249,61 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
         rightExpressionSymbol.ifPresent(symbol -> result.addSymbolStatistics(symbol, rightNullsFiltered));
         return result.build();
     }
+
+    private static PlanNodeStatsEstimate estimateExpressionToExpressionInequality(
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:254]: I've updated the calculation to be a bit more sophisticated now. But this case will require guesswork.
Even with the usual uniform distribution assumption, we wouldn't know about the selectivity of the predicate within the overlapping part of both ranges unless we have some understanding of the correlation between the values of both columns.

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java (hunk 2 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java]: extracting to a method results in a long list of arguments to the methods, so keeping it as is for now

**Comment by sopel39** [comment]
> The inequality estimation code path is executed only for q72 in tpcds and there was no change to plan there.

You mean partitioned q72 plan?

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 7]
```diff
@@ -239,4 +250,93 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                }
+                // left is always greater than right, no overlap
+                if (leftRange.getLow() > rightRange.getHigh()) {
+                    return inputStatistics.mapOutputRowCount(rowCount -> 0.0);
+                }
+                if (isNonEstimatablePredicateApproximationEnabled) {
+                    double overlappingRangeFraction = leftRange.overlapPercentWith(leftRange.intersect(rightRange));
+                    double alwaysLessRangeFraction = 0.0;
+                    if (leftRange.getLow() < rightRange.getLow()) {
+                        alwaysLessRangeFraction = leftRange.overlapPercentWith(new StatisticRange(leftRange.getLow(), rightRange.getLow(), NaN));
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:305]: I don't think intersect works well `NaN` NDVs. 

Why not to do `1.0 - leftRange.overlapPercentWith(rightRange)`?

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 8]
```diff
@@ -239,4 +250,93 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                    }
+                    double filterFactor = alwaysLessRangeFraction + (UNKNOWN_FILTER_COEFFICIENT * overlappingRangeFraction);
+                    return inputStatistics.mapOutputRowCount(rowCount -> rowCount * filterFactor * nullsFilterFactor);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:308]: You can derive symbol ranges:
I think it is
left: `[current_left_min, min(current_left_max, current_right_max)]`
right: `[max(current_right_min, current_left_min), current_right_max]`
You should adjust NDVs of symbols accordingly

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 9]
```diff
@@ -239,4 +250,93 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                            symbol,
+                            rightExpressionStatistics.mapNullsFraction(nullsFraction -> 0.0)));
+                    return estimate.setOutputRowCount(inputStatistics.getOutputRowCount() * nullsFilterFactor)
+                            .build();
+                }
+                // left is always greater than right, no overlap
+                if (leftRange.getLow() > rightRange.getHigh()) {
+                    return inputStatistics.mapOutputRowCount(rowCount -> 0.0);
+                }
+                if (isNonEstimatablePredicateApproximationEnabled) {
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:303]: That causes variable names to start conflicting between the switch cases due to the increased scope

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java (hunk 7 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:308]: That would include the part of `leftRange` that is greater than `rightRange` (e.g. x = [-10, 10], y = [0, 5]), we want to avoid that.

**Comment by raunaqmorarka** [comment]
Updated benchmarks
[Inequality estimation 2 sf1000 orc partitioned.pdf](https://github.com/trinodb/trino/files/8327057/Inequality.estimation.2.sf1000.orc.partitioned.pdf)
[Inequality estimation 2 sf1000 orc unpartitioned.pdf](https://github.com/trinodb/trino/files/8327058/Inequality.estimation.2.sf1000.orc.unpartitioned.pdf)

Ignore results from all the TPCDS queries other than q72 as the inequality estimation code path is encountered only in that TPCDS query.

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 10]
```diff
@@ -22,7 +22,9 @@
 import static io.trino.cost.SymbolStatsEstimate.buildFrom;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:22]: I don't think it needs to be behind flag. We actually added pretty reasonable estimate.

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 11]
```diff
@@ -239,4 +252,134 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+        StatisticRange leftRange = StatisticRange.from(leftExpressionStatistics);
+        StatisticRange rightRange = StatisticRange.from(rightExpressionStatistics);
+        if (leftRange.isEmpty() || rightRange.isEmpty()) {
+            return inputStatistics.mapOutputRowCount(rowCount -> 0.0);
+        }
+        // We don't know the correlation between NULLs, so we take the max nullsFraction from the expression statistics
+        // to make a conservative estimate for nullsFilterFactor
+        double nullsFilterFactor;
+        if (isNaN(leftExpressionStatistics.getNullsFraction()) && isNaN(rightExpressionStatistics.getNullsFraction())) {
+            nullsFilterFactor = 1.0;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:278]: nit: I think we might just bail in this case, but I'm fine leaving it too

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 12]
```diff
@@ -239,4 +252,134 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+            return PlanNodeStatsEstimate.unknown();
+        }
+
+        StatisticRange leftRange = StatisticRange.from(leftExpressionStatistics);
+        StatisticRange rightRange = StatisticRange.from(rightExpressionStatistics);
+        if (leftRange.isEmpty() || rightRange.isEmpty()) {
+            return inputStatistics.mapOutputRowCount(rowCount -> 0.0);
+        }
+        // We don't know the correlation between NULLs, so we take the max nullsFraction from the expression statistics
+        // to make a conservative estimate for nullsFilterFactor
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:275]: nit: I would add that we assume that `nulls` are fully correlated, e.g:
```
conservative estimate (nulls are fully correlated)
```

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 13]
```diff
@@ -239,4 +252,134 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+            nullsFilterFactor = 1.0;
+        }
+        else {
+            nullsFilterFactor = 1 - maxExcludeNaN(leftExpressionStatistics.getNullsFraction(), rightExpressionStatistics.getNullsFraction());
+        }
+
+        switch (operator) {
+            case LESS_THAN:
+            case LESS_THAN_OR_EQUAL:
+                // left is always greater than right, no overlap
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:287]: do we actually need these special branches
```
1. if (leftRange.getLow() > rightRange.getHigh())
2. if (leftRange.getHigh() < rightRange.getLow()) {
```

File: core/trino-main/src/test/java/io/trino/cost/TestFilterStatsCalculator.java [hunk 14]
```diff
@@ -205,6 +207,123 @@ public void testComparison()
         }
     }
 
+    @Test
+    public void testInequalityComparisons()
+    {
+        double inputRowCount = standardInputStatistics.getOutputRowCount();
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/cost/TestFilterStatsCalculator.java:213]: please add some test cases with `NaN` and `Infinites` (min, max, null fraction, Ndv)

**Review by sopel39 — COMMENTED** [review]
lgtm % comments

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 15]
```diff
@@ -239,4 +252,134 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                        alwaysLessRangeFraction = leftRange.overlapPercentWith(new StatisticRange(leftRange.getLow(), rightRange.getLow(), NaN));
+                    }
+                    double filterFactor = alwaysLessRangeFraction + (OVERLAPPING_RANGE_INEQUALITY_FILTER_COEFFICIENT * overlappingRangeFraction);
+                    leftExpressionSymbol.ifPresent(symbol -> estimate.addSymbolStatistics(
+                            symbol,
+                            SymbolStatsEstimate.builder()
+                                    .setLowValue(leftRange.getLow())
+                                    .setHighValue(minExcludeNaN(leftRange.getHigh(), rightRange.getHigh()))
+                                    .setAverageRowSize(leftExpressionStatistics.getAverageRowSize())
+                                    .setDistinctValuesCount(leftExpressionStatistics.getDistinctValuesCount() * filterFactor)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:317]: I don't think NDV count is reduced apart from `alwaysGreaterRange`
1. in `alwaysLessRange` no NDVs are filtered
2. in overlapping range, each NDV should be part of at least one pair

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java (hunk 11 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java]: updated to bail in this case

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java (hunk 10 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:22]: I would prefer to keep the flag for now and remove it after a few releases once we have validation from real workloads.

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java (hunk 13 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:284]: We can simplify that when we remove the flag

File: core/trino-main/src/test/java/io/trino/cost/TestFilterStatsCalculator.java (hunk 14 above)
Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/cost/TestFilterStatsCalculator.java:212]: I've added tests to TestComparisonStatsCalculator

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 16]
```diff
@@ -239,4 +252,163 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                            symbol,
+                            SymbolStatsEstimate.builder()
+                                    .setLowValue(leftRange.getLow())
+                                    .setHighValue(minExcludeNaN(leftRange.getHigh(), rightRange.getHigh()))
+                                    .setAverageRowSize(leftExpressionStatistics.getAverageRowSize())
+                                    .setDistinctValuesCount(leftExpressionStatistics.getDistinctValuesCount() * (leftAlwaysLessRangeFraction + leftOverlappingRangeFraction))
+                                    .setNullsFraction(0)
+                                    .build()));
+
+                    double rightOverlappingRangeFraction = rightRange.overlapPercentWith(intersection);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:325]: `rightRange.overlapPercentWith(leftRange);`

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 17]
```diff
@@ -239,4 +252,163 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                if (isNonEstimatablePredicateApproximationEnabled) {
+                    PlanNodeStatsEstimate.Builder estimate = PlanNodeStatsEstimate.buildFrom(inputStatistics);
+                    StatisticRange intersection = leftRange.intersect(rightRange);
+
+                    double leftOverlappingRangeFraction = leftRange.overlapPercentWith(intersection);
+                    double leftAlwaysLessRangeFraction;
+                    if (leftRange.getLow() < rightRange.getLow()) {
+                        leftAlwaysLessRangeFraction = min(
+                                leftRange.overlapPercentWith(new StatisticRange(leftRange.getLow(), rightRange.getLow(), NaN)),
+                                1 - leftOverlappingRangeFraction);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:309]: add a comment when it's needed

**Review by sopel39 — COMMENTED** [review]
lgtm % comments

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java (hunk 10 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:22]: removed flag

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 18]
```diff
@@ -239,4 +252,163 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                                    .setAverageRowSize(rightExpressionStatistics.getAverageRowSize())
+                                    .setDistinctValuesCount(rightExpressionStatistics.getDistinctValuesCount() * (rightOverlappingRangeFraction + rightAlwaysGreaterRangeFraction))
+                                    .setNullsFraction(0)
+                                    .build()));
+                    return estimate.setOutputRowCount(inputStatistics.getOutputRowCount() * nullsFilterFactor * filterFactor)
+                            .build();
+                }
+                return PlanNodeStatsEstimate.unknown();
+            case GREATER_THAN:
+            case GREATER_THAN_OR_EQUAL:
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:289]: did that now
note that I had to tweak filterFactor calculation a bit to make it symmetrical

File: core/trino-main/src/test/java/io/trino/cost/TestComparisonStatsCalculator.java [hunk 19]
```diff
@@ -695,6 +708,237 @@ public void symbolToCastExpressionNotEqual()
...
+                        .distinctValuesCount(40)
+                        .nullsFraction(0))
+                .symbolStats("w", symbolAssert -> symbolAssert.averageRowSize(8)
+                        .lowValue(0)
+                        .highValue(20)
+                        .distinctValuesCount(30)
+                        .nullsFraction(0))
+                .symbolStats("z", equalTo(capNDV(zStats, rowCount)));
+
+        // Open ranges
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/cost/TestComparisonStatsCalculator.java:892]: do we have case for `unknowns` (not open ranges)?

File: core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java [hunk 20]
```diff
@@ -239,4 +253,124 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
...
+                    rightExpressionStatistics.mapNullsFraction(nullsFraction -> 0.0)));
+            return estimate.setOutputRowCount(inputStatistics.getOutputRowCount() * nullsFilterFactor)
+                    .build();
+        }
+
+        PlanNodeStatsEstimate.Builder estimate = PlanNodeStatsEstimate.buildFrom(inputStatistics);
+        double leftOverlappingRangeFraction = leftRange.overlapPercentWith(rightRange);
+        double leftAlwaysLessRangeFraction;
+        if (leftRange.getLow() < rightRange.getLow()) {
+            leftAlwaysLessRangeFraction = min(
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java:333]: this is wrong. For
```
left: [0, 3]
right: [0, 6]
```
pairs of:
`left: [0,3] right:[3, 6]` should be part of `leftAlwaysLessRangeFraction`

File: core/trino-main/src/test/java/io/trino/cost/TestComparisonStatsCalculator.java (hunk 19 above)
Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/cost/TestComparisonStatsCalculator.java:888]: we have a case with `unknownRange` and `unknownNdvRange`

**Review by sopel39 — APPROVED** [review]
mind automation (if it fails)
